File: VideoApp/VideoApp/Video/VideoAppCameraSource.m
//
//  Copyright (C) 2019 Twilio, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "VideoAppCameraSource.h"
#import "LocalMediaController+Internal.h"
#import "TVILocalVideoTrack+Mirroring.h"
#import "VideoApp-Swift.h"

// 640x480 squarish crop (1.13:1)
static const CMVideoDimensions kVideoAppCameraSourceDimensions = (CMVideoDimensions){544, 480};
// 1024x768 squarish crop (1.25:1) on most iPhones. 1280x720 squarish crop (1.25:1) on the iPhone X and models that don't have 1024x768.
static const CMVideoDimensions kVideoAppCameraSourceSimulcastDimensions = (CMVideoDimensions){900, 720};

static const int32_t kVideoAppCameraSourceFrameRate = 20;
// With simulcast enabled there are 3 temporal layers, allowing a frame rate of f/4.
static const int32_t kVideoAppCameraSourceSimulcastFrameRate = 24;

TVIVideoFormat *VideoAppCameraSourceSelectVideoFormatBySize(AVCaptureDevice *device, CMVideoDimensions targetSize) {
    TVIVideoFormat *selectedFormat = nil;
    // Ordered from smallest to largest.
    NSOrderedSet<TVIVideoFormat *> *formats = [TVICameraSource supportedFormatsForDevice:device];

    for (TVIVideoFormat *format in formats) {
        if (format.pixelFormat != TVIPixelFormatYUV420BiPlanarFullRange) {
            continue;
        }
        CMVideoDimensions dimensions = format.dimensions;

        // Cropping might be used if there is not an exact match.
        if (dimensions.width >= targetSize.width && dimensions.height >= targetSize.height) {
            selectedFormat = format;
            break;
        }
    }
    return selectedFormat;
}

@interface VideoAppCameraSource () <TVICameraSourceDelegate>

@property (nonatomic, strong) TVICameraSource *cameraSource;
@property (nonatomic, strong) TVILocalVideoTrack *localVideoTrack;
@property (nonatomic, weak) LocalMediaController *localMediaController;

@end

@implementation VideoAppCameraSource

- (instancetype)initWithLocalMediaController:(LocalMediaController *)localMediaController {
    self = [super init];

    if (self != nil) {
        _localMediaController = localMediaController;
    }

    return self;
}

- (void)dealloc {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    // Prevent leaking the CameraSource in case a Track/Source has been created.
    [_cameraSource stopCapture];
}

- (TVILocalVideoTrack *)localVideoTrack {
    if (_localVideoTrack == nil) {
        [self createCameraSource];
    }

    return _localVideoTrack;
}

- (void)createCameraSource {
    TVICameraSourceOptions *options = [TVICameraSourceOptions optionsWithBlock:^(TVICameraSourceOptionsBuilder * _Nonnull builder) {
        if (@available(iOS 13, *)) {
            UIWindowScene *keyScene = [UIApplication sharedApplication].keyWindow.windowScene;
            builder.orientationTracker = [TVIUserInterfaceTracker trackerWithScene:keyScene];
        }
        // If a Group Room is being requested, take a best guess and remove rotation tags using hardware acceleration.
        if ([SwiftToObjc isGroupTopology]) {
            builder.rotationTags = TVICameraSourceOptionsRotationTagsRemove;
        }
    }];
    self.cameraSource = [[TVICameraSource alloc] initWithOptions:options delegate:self];

    if (self.cameraSource == nil) {
        NSLog(@"Unable to create a capturer...");
        return;
    }

    self.localVideoTrack = [TVILocalVideoTrack trackWithSource:self.cameraSource
                                                       enabled:YES
                                                          name:@"camera"];

    [self startCameraSourceWithPosition:AVCaptureDevicePositionFront];
}

- (void)startCameraSourceWithPosition:(AVCaptureDevicePosition)position {
    // Find a format and start capture.
    AVCaptureDevice *camera = [TVICameraSource captureDeviceForPosition:position];
    CMVideoDimensions dimensions;
    CGFloat cropRatio;
    int32_t frameRate;
    if ([SwiftToObjc enableVP8Simulcast]) {
        cropRatio = (CGFloat)kVideoAppCameraSourceSimulcastDimensions.width / (CGFloat)kVideoAppCameraSourceSimulcastDimensions.height;
        dimensions = kVideoAppCameraSourceSimulcastDimensions;
        frameRate = kVideoAppCameraSourceSimulcastFrameRate;
    } else {
        cropRatio = (CGFloat)kVideoAppCameraSourceDimensions.width / (CGFloat)kVideoAppCameraSourceDimensions.height;
        dimensions = kVideoAppCameraSourceDimensions;
        frameRate = kVideoAppCameraSourceFrameRate;
    }
    TVIVideoFormat *preferredFormat = VideoAppCameraSourceSelectVideoFormatBySize(camera, dimensions);
    preferredFormat.frameRate = MIN(preferredFormat.frameRate, frameRate);

    CMVideoDimensions cropDimensions = preferredFormat.dimensions;
    if (cropDimensions.width > cropDimensions.height) {
        cropDimensions.width = cropDimensions.height * cropRatio;
    } else {
        cropDimensions.height = cropDimensions.width * cropRatio;
    }
    
    TVIVideoFormat *outputFormat = [[TVIVideoFormat alloc] init];
    outputFormat.dimensions = cropDimensions;
    outputFormat.pixelFormat = preferredFormat.pixelFormat;
    outputFormat.frameRate = 0;

    NSLog(@"Cropping camera output to format: %@", outputFormat);

    [self.cameraSource requestOutputFormat:outputFormat];

    typeof(self) __weak weakSelf = self;
    [self.cameraSource startCaptureWithDevice:camera format:preferredFormat completion:^(AVCaptureDevice *device,
                                                                                              TVIVideoFormat *startFormat,
                                                                                              NSError *error) {
        if (!error) {
            [weakSelf.localMediaController videoCaptureStarted];
        }
    }];
}

- (void)destroyLocalVideoTrack {
    [self.cameraSource stopCaptureWithCompletion:^(NSError *error) {
        self.localVideoTrack = nil;
        self.cameraSource = nil;
    }];
}

- (void)flipCamera {
    AVCaptureDevicePosition position = self.cameraSource.device.position;
    AVCaptureDevicePosition nextPosition = position == AVCaptureDevicePositionFront ? AVCaptureDevicePositionBack : AVCaptureDevicePositionFront;

    typeof(self) __weak weakSelf = self;
    [self.cameraSource selectCaptureDevice:[TVICameraSource captureDeviceForPosition:nextPosition] completion:^(AVCaptureDevice *device, TVIVideoFormat *format, NSError *error) {
        if (!error) {
            [weakSelf.localMediaController videoCaptureStarted];
        }
    }];
}

- (BOOL)shouldMirrorLocalVideoView {
    return [self.localVideoTrack tvi_shouldMirror];
}

#pragma mark - TVICameraSourceDelegate

- (void)cameraSourceInterruptionEnded:(TVICameraSource *)source {
    [self.localMediaController cameraSourceInterruptionEnded];
}

- (void)cameraSourceWasInterrupted:(TVICameraSource *)source reason:(AVCaptureSessionInterruptionReason)reason {
    [self.localMediaController cameraSourceWasInterrupted];
}

- (void)cameraSource:(TVICameraSource *)source didFailWithError:(NSError *)error {

}

@end

File: VideoApp/VideoApp/Video/VideoAppCameraSource.swift
//
//  Copyright (C) 2020 Twilio, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import TwilioVideo

@objc class VideoAppCameraSource: NSObject, VideoAppCamera {
    private(set) var localVideoTrack: LocalVideoTrack?
    private let appSettingsStore: AppSettingsStoreWriting = AppSettingsStore.shared
    private weak var localMediaController: LocalMediaController?
    private var cameraSource: CameraSource?

    @objc init(localMediaController: LocalMediaController) {
        self.localMediaController = localMediaController
        super.init()
        createCameraSource()
    }
    
    deinit {
        cameraSource?.stopCapture() // Prevent leaking the CameraSource when a Track/Source has been created
    }

    func destroyLocalVideoTrack() {
        cameraSource?.stopCapture() { [weak self] _ in
            self?.localVideoTrack = nil
            self?.cameraSource = nil
        }
    }
    
    func flip() {
        guard
            let cameraSource = cameraSource,
            let position = cameraSource.device?.position,
            let captureDevice = CameraSource.captureDevice(position: position == .front ? .back : .front)
            else {
                return
        }
        
        cameraSource.selectCaptureDevice(captureDevice) { [weak self] _, _, error in
            guard error == nil else { return }
            
            self?.localMediaController?.videoCaptureStarted()
        }
    }

    func shouldMirrorLocalVideoView() -> Bool {
        localVideoTrack?.shouldMirror ?? false
    }

    private func createCameraSource() {
        let options = CameraSourceOptions() { builder in
            if #available(iOS 13, *) {
                builder.orientationTracker = UserInterfaceTracker(scene: UIApplication.shared.keyWindow!.windowScene!)
            }
            
            switch self.appSettingsStore.topology {
            case .group:
                // Take a best guess and remove rotation tags using hardware acceleration
                builder.rotationTags = .remove
            case .peerToPeer:
                break
            }
        }
        
        cameraSource = CameraSource(options: options, delegate: self)
        
        if let cameraSource = cameraSource {
            localVideoTrack = LocalVideoTrack(source: cameraSource, enabled: true, name: "camera")
            startCameraSource()
        } else {
            print("Unable to create a capturer...")
        }
    }
    
    private func startCameraSource() {
        guard
            let cameraSource = cameraSource,
            let captureDevice = CameraSource.captureDevice(position: .front)
            else {
                return
        }

        let targetSize: CMVideoDimensions
        let cropRatio: CGFloat
        let frameRate: UInt
        
        switch appSettingsStore.videoCodec {
        case .h264, .vp8:
            // 640 x 480 squarish crop (1.13:1)
            targetSize = CMVideoDimensions(width: 544, height: 480)
            
            cropRatio = CGFloat(targetSize.width) / CGFloat(targetSize.height)
            frameRate = 20
        case .vp8Simulcast:
            // 1024 x 768 squarish crop (1.25:1) on most iPhones. 1280 x 720 squarish crop (1.25:1) on the iPhone X
            // and models that don't have 1024 x 768.
            targetSize = CMVideoDimensions(width: 900, height: 720)
            
            cropRatio = CGFloat(targetSize.width) / CGFloat(targetSize.height)
            frameRate = 24 // With simulcast enabled there are 3 temporal layers, allowing a frame rate of f/4
        }
        
        let preferredFormat = selectVideoFormatBySize(captureDevice: captureDevice, targetSize: targetSize)
        preferredFormat.frameRate = min(preferredFormat.frameRate, frameRate)
        
        let cropDimensions: CMVideoDimensions
        
        if preferredFormat.dimensions.width > preferredFormat.dimensions.height {
            cropDimensions = CMVideoDimensions(
                width: Int32(CGFloat(preferredFormat.dimensions.height) * cropRatio),
                height: preferredFormat.dimensions.height
            )
        } else {
            cropDimensions = CMVideoDimensions(
                width: preferredFormat.dimensions.width,
                height: Int32(CGFloat(preferredFormat.dimensions.width) * cropRatio)
            )
        }
        
        let outputFormat = VideoFormat()
        outputFormat.dimensions = cropDimensions
        outputFormat.pixelFormat = preferredFormat.pixelFormat
        outputFormat.frameRate = 0
        
        cameraSource.requestOutputFormat(outputFormat)
        
        cameraSource.startCapture(device: captureDevice, format: preferredFormat) { [weak self] _, _, error in
            guard error == nil else { return }
            
            self?.localMediaController?.videoCaptureStarted()
        }
    }
        
    private func selectVideoFormatBySize(captureDevice: AVCaptureDevice, targetSize: CMVideoDimensions) -> VideoFormat {
        let supportedFormats = Array(CameraSource.supportedFormats(captureDevice: captureDevice)) as! [VideoFormat]
        
        // Cropping might be used if there is not an exact match
        for format in supportedFormats {
            guard
                format.pixelFormat == .formatYUV420BiPlanarFullRange &&
                    format.dimensions.width >= targetSize.width &&
                    format.dimensions.height >= targetSize.height
                else {
                    continue
            }
            
            return format
        }
        
        fatalError()
    }
}

extension VideoAppCameraSource: CameraSourceDelegate {
    func cameraSourceInterruptionEnded(source: CameraSource) {
        localMediaController?.cameraSourceInterruptionEnded()
    }

    func cameraSourceWasInterrupted(source: CameraSource, reason: AVCaptureSession.InterruptionReason) {
        localMediaController?.cameraSourceWasInterrupted()
    }
}
